cgen/sycl.py

```diff
@@ -0,0 +1,240 @@
+    elif dtype == np.int32:
+        return "int"
+    elif dtype == np.uint32:
+        return "uint"
+    elif dtype == np.int16:
+        return "short"
+    elif dtype == np.uint16:
+        return "ushort"
+    elif dtype == np.int8:
+        return "char"
+    elif dtype == np.uint8:
+        return "uchar"
+    elif dtype == np.float32:
+        return "float"
+    elif dtype == np.float64:
+        return "double"
+    else:
+        raise ValueError(f"unable to map dtype '{dtype}'")
+
+
+# {{{ kernel
+
+
+class SYCLKernel(DeclSpecifier):
+    def __init__(self, subdecl, ndim, ndrange, queue):
+        self.ndim = ndim
+        subdecl.arg_decls.append(Value("sycl::queue", queue))
+        subdecl.arg_decls.append(
+            Value("sycl::nd_range<{}> ".format(self.ndim), ndrange)
+        )
+        DeclSpecifier.__init__(self, subdecl, "")
+
+    mapper_method = "map_sycl_kernel"
+
+
+# }}}
+
+
+# {{{ kernel args
+
+
+class SYCLConstant(DeclSpecifier):
+    def __init__(self, subdecl):
+        self.subdecl = subdecl
+
+    def get_decl_pair(self):
+        sub_tp, sub_decl = self.subdecl.get_decl_pair()
+        return [f"sycl::constant_ptr<{sub_tp[0]}>"], sub_decl
+
+    mapper_method = "map_sycl_constant"
+
+
+class SYCLLocal(NestedDeclarator):
+    def __init__(self, subdecl, item):
+        self.item = item
+        self.subdecl = subdecl
+
+    def get_decl_pair(self):
+        sub_tp, sub_decl = self.subdecl.get_decl_pair()
+        if isinstance(self.subdecl, ArrayOf):
+            return ["auto&"], Assign(
+                self.subdecl.get_array_name(),
+                "*sycl::ext::oneapi::group_local_memory"
+                f"<{sub_tp[0]}[{self.subdecl.get_size()}]>"
+                f"({self.item}.get_group())")
+        else:
+            return ["auto&"], Assign(
+                sub_decl,
+                f"*sycl::ext::oneapi::group_local_memory<{sub_tp[0]}>"
+                f"({self.item}.get_group())")
+
+    mapper_method = "map_sycl_local"
+
+
+class SYCLGlobal(DeclSpecifier):
+    def __init__(self, subdecl):
+        self.subdecl = subdecl
+
+    def get_decl_pair(self):
+        sub_tp, sub_decl = self.subdecl.get_decl_pair()
+        return [f"sycl::global_ptr<{sub_tp[0]}>"], sub_decl
+
+    mapper_method = "map_sycl_global"
+
+
+# TODO sycl Image
+class SYCLImage(Value):
+    def __init__(self, dims, mode, name):
+        if mode == "r":
+            spec = "__read_only"
+        elif mode == "w":
+            spec = "__write_only"
+        else:
+            raise ValueError("mode must be one of 'r' or 'w'")
+
+        Value.__init__(self, f"{spec} image{dims}d_t", name)
+
+    mapper_method = "map_cl_image"
+
+
+# }}}
+
+
+# {{{ function attributes
+class _SYCLWorkGroupSizeDeclarator(Generable):
+    def __init__(self, dim):
+
+        while len(dim) < 3:
+            dim = dim + (1,)
+        self.dim = dim
+        self.decl = 'extern "C" [[sycl::reqd_work_group_size({})]]'.format(
```

Comment:
Is the `extern "C"` necessarily part of this? It would seem that this prohibits overloading.